src/java.base/share/classes/java/lang/Float16.java

```diff
@@ -859,36 +873,188 @@ public static Float16 sqrt(Float16 radicand) {
+         * Float16 values less than the precision width of double.
+         *
+         * In cases where the numerical value of (a * b) + c is
+         * computed exactly in a double, after a single rounding to
+         * Float16, the result is of necessity correct since the one
+         * double -> Float16 conversion is the only source of
+         * numerical error. The operation as implemented in those
+         * cases would be equivalent to rounding the infinitely precise
+         * value to the result format, etc.
+         *
+         * However, for some inputs, the intermediate product-sum will
+         * *not* be exact and additional analysis is needed to justify
+         * not having any corrective computation to compensate for
+         * intermediate rounding errors.
+         *
+         * The following analysis will rely on the range of bit
+         * positions representable in the intermediate
+         * product-sum.
+         *
+         * For the product a*b of Float16 inputs, the range of
+         * exponents for nonzero finite results goes from 2^(-28)
+         * (from MIN_VALUE squared) to 2^31 (from the exact value of
+         * MAX_VALUE squared). This full range of exponent positions,
+         * (31 -(-28) + 1 ) = 60 exceeds the precision of
+         * double. However, only the product a*b can exceed the
+         * exponent range of Float16. Therefore, there are three main
+         * cases to consider:
+         *
+         * 1) Large exponent product, exponent > Float16.MAX_EXPONENT
+         *
+         * The magnitude of the overflow threshold for Float16 is:
+         *
+         * MAX_VALUE + 1/2 * ulp(MAX_VALUE) =  0x1.ffcp15 + 0x0.002p15 = 0x1.ffep15
+         *
+         * Therefore, any product greater than 0x1.ffep15 + MAX_VALUE
+         * = 0x1.ffdp16 will certainly overflow (under round to
+         * nearest) since adding in c = -MAX_VALUE will still be above
+         * the overflow threshold.
+         *
+         * If the exponent of the product is 15 or 16, the smallest
+         * subnormal Float16 is 2^-24 and the ~40 bit wide range bit
+         * positions would fit in a single double exactly.
+         *
+         * 2) Exponent of product is within the range of _normal_
+         * Float16 values; Float16.MIN_EXPONENT <=  exponent <= Float16.MAX_EXPONENT
+         *
+         * The exact product has at most 22 contiguous bits in its
+         * logical significand. The third number being added in has
+         * at most 11 contiguous bits in its significand and the
+         * lowest bit position that could be set is
+         * 2^(-24). Therefore, when the product has the maximum
+         * in-range exponent, 2^15, a single double has enough
+         * precision to hold down to the smallest subnormal bit
+         * position, 15 - (-24) + 1 = 40 < 53. If the product was
+         * large and overflowed when the third operand was added, this
```

Comment:
In `double` arithmetic, as here, `a * b + c` cannot overflow, so this is a bit confusing.